File: sample/py/03_julia_set.py
# 03: ジュリア集合を PNG 形式で出力するサンプルです。
# トランスパイル互換を意識し、単純なループ中心で実装しています。

from time import perf_counter
import png_helper


def render_julia(width: int, height: int, max_iter: int, cx: float, cy: float) -> bytearray:
    pixels: bytearray = bytearray()

    y: int = 0
    while y < height:
        zy0: float = -1.2 + 2.4 * ((y * 1.0) / ((height - 1) * 1.0))

        x: int = 0
        while x < width:
            zx: float = -1.8 + 3.6 * ((x * 1.0) / ((width - 1) * 1.0))
            zy: float = zy0

            i: int = 0
            while i < max_iter:
                zx2: float = zx * zx
                zy2: float = zy * zy
                if zx2 + zy2 > 4.0:
                    break
                zy = 2.0 * zx * zy + cy
                zx = zx2 - zy2 + cx
                i = i + 1

            r: int = 0
            g: int = 0
            b: int = 0
            if i >= max_iter:
                r = 0
                g = 0
                b = 0
            else:
                t: float = (i * 1.0) / (max_iter * 1.0)
                r = int(255.0 * (0.2 + 0.8 * t))
                g = int(255.0 * (0.1 + 0.9 * (t * t)))
                b = int(255.0 * (1.0 - t))

            pixels.append(r)
            pixels.append(g)
            pixels.append(b)
            x = x + 1

        y = y + 1

    return pixels


def run_julia() -> None:
    width: int = 1280
    height: int = 720
    max_iter: int = 520
    out_path: str = "sample/out/julia_03.png"

    start: float = perf_counter()
    pixels: bytearray = render_julia(width, height, max_iter, -0.8, 0.156)
    png_helper.write_rgb_png(out_path, width, height, pixels)
    elapsed: float = perf_counter() - start

    print("output:", out_path)
    print("size:", width, "x", height)
    print("max_iter:", max_iter)
    print("elapsed_sec:", elapsed)


if __name__ == "__main__":
    run_julia()

File: sample/cs/03_julia_set.cs
using System.Collections.Generic;
using System.IO;
using System;

public static class Program
{
    public static List<byte> render_julia(long width, long height, long max_iter, double cx, double cy)
    {
        List<byte> pixels = new List<byte>();
        var __pytra_range_start_1 = 0;
        var __pytra_range_stop_2 = height;
        var __pytra_range_step_3 = 1;
        if (__pytra_range_step_3 == 0) throw new Exception("range() arg 3 must not be zero");
        for (var y = __pytra_range_start_1; (__pytra_range_step_3 > 0) ? (y < __pytra_range_stop_2) : (y > __pytra_range_stop_2); y += __pytra_range_step_3)
        {
            double zy0 = ((-1.2) + (2.4 * ((double)(y) / (double)((height - 1L)))));
            var __pytra_range_start_4 = 0;
            var __pytra_range_stop_5 = width;
            var __pytra_range_step_6 = 1;
            if (__pytra_range_step_6 == 0) throw new Exception("range() arg 3 must not be zero");
            for (var x = __pytra_range_start_4; (__pytra_range_step_6 > 0) ? (x < __pytra_range_stop_5) : (x > __pytra_range_stop_5); x += __pytra_range_step_6)
            {
                double zx = ((-1.8) + (3.6 * ((double)(x) / (double)((width - 1L)))));
                double zy = zy0;
                long i = 0L;
                while (Pytra.CsModule.py_runtime.py_bool((i < max_iter)))
                {
                    double zx2 = (zx * zx);
                    double zy2 = (zy * zy);
                    if (Pytra.CsModule.py_runtime.py_bool(((zx2 + zy2) > 4.0)))
                    {
                        break;
                    }
                    zy = (((2.0 * zx) * zy) + cy);
                    zx = ((zx2 - zy2) + cx);
                    i = (i + 1L);
                }
                long r = 0L;
                long g = 0L;
                long b = 0L;
                if (Pytra.CsModule.py_runtime.py_bool((i >= max_iter)))
                {
                    r = 0L;
                    g = 0L;
                    b = 0L;
                }
                else
                {
                    double t = ((double)(i) / (double)(max_iter));
                    r = (long)((255.0 * (0.2 + (0.8 * t))));
                    g = (long)((255.0 * (0.1 + (0.9 * (t * t)))));
                    b = (long)((255.0 * (1.0 - t)));
                }
                Pytra.CsModule.py_runtime.py_append(pixels, r);
                Pytra.CsModule.py_runtime.py_append(pixels, g);
                Pytra.CsModule.py_runtime.py_append(pixels, b);
            }
        }
        return pixels;
    }

    public static void run_julia()
    {
        long width = 3840L;
        long height = 2160L;
        long max_iter = 20000L;
        string out_path = "sample/out/julia_03.png";
        double start = Pytra.CsModule.time.perf_counter();
        List<byte> pixels = render_julia(width, height, max_iter, (-0.8), 0.156);
        Pytra.CsModule.png_helper.write_rgb_png(out_path, width, height, pixels);
        double elapsed = (Pytra.CsModule.time.perf_counter() - start);
        Pytra.CsModule.py_runtime.print("output:", out_path);
        Pytra.CsModule.py_runtime.print("size:", width, "x", height);
        Pytra.CsModule.py_runtime.print("max_iter:", max_iter);
        Pytra.CsModule.py_runtime.print("elapsed_sec:", elapsed);
    }

    public static void Main(string[] args)
    {
        run_julia();
    }
}
